File: rate_limiter/token_bucket/token_bucketA.js
//bucket con capacidad de tokens
//rate de generador de tokens
//request
//process

let bucket = []
let bucketCapacity = 5
let rate = 2000 //ms



function generateRequests() {
    let requests = Array.from({ length: 5 }, () => `req_${Math.floor(Math.random() * 100)}`)
    return requests
}

async function processRequests() {
    return new Promise(resolve => {
        if (bucket.length > 0) {
            setTimeout(() => {
                bucket.shift()
                console.log('request processed', bucket)
                resolve()
            }, 1000)
        }
        else {
            console.log('no tokens available')
            resolve()
        }
    })
}

function generateTokens() {
    if (bucket.length == 5) return
    if (bucket.length == (bucketCapacity - 1)) {
        //generate 1
        bucket.push('token')
        return
    }
    bucket.push('token', 'token')
}


async function main() {
    setInterval(() => {
        generateTokens()
    }, 1000)
    setInterval(() => {
        processRequests().then(() => {
            console.log(bucket)
        })

    }, 1000)

}

await main()

File: rate_limiter/token_bucket/token_bucketA.py
import asyncio




bucket = []
bucket_capacity = 5
rate = 2000 #ms


async def process_requests():
    if len(bucket) == 0:
        return
    await asyncio.sleep(1)
    del bucket[:1]
    print('request processed', bucket)
    return


def generate_tokens():
    if len(bucket) == bucket_capacity:
        return
    if len(bucket) == (bucket_capacity - 1):
        bucket.append('token')
        return
    bucket.append('token')
    bucket.append('token')
    
    

async def main():
    while(True):
        generate_tokens()
        await process_requests()

asyncio.run(main())